File: verification/insert.c
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

struct node {
	int key;
	struct node *left, *right;
};

// A utility function to create a new BST node
struct node* newNode(int item)
{
	struct node* temp
		= (struct node*)malloc(sizeof(struct node));
	temp->key = item;
	temp->left = NULL;
	temp->right = NULL;
	return temp;
}

uint8_t* insert_c(uint8_t* nodeptr, int key)
{
	struct node* node = (struct node*)nodeptr;
	/* If the tree is empty, return a new node */
	if (node == NULL)
		return (uint8_t*) newNode(key);

	/* Otherwise, recur down the tree */
	if (key < node->key)
		node->left = (struct node*) insert_c( (uint8_t*) node->left, key);
	else
		node->right = (struct node*) insert_c((uint8_t*) node->right, key);

	/* return the (unchanged) node pointer */
	return (uint8_t*) node;
}

void main(){
	
}

File: verification/insert.rs
use std::ptr;

#[repr(C)]
struct Node {
    key: i32,
    left: *mut Node,
    right: *mut Node,
}

// A utility function to create a new BST node
unsafe fn new_node(item: i32) -> *mut Node {
    let temp = Box::into_raw(Box::new(Node {
        key: item,
        left: ptr::null_mut(),
        right: ptr::null_mut(),
    }));
    temp
}

#[no_mangle]
fn insert_rust(nodeptr: *mut u8, key: i32) -> *mut u8 {
    unsafe {
        let node = nodeptr as *mut Node;

        // If the tree is empty, return a new node
        if node.is_null() {
            return new_node(key) as *mut u8;
        }

        // Otherwise, recur down the tree
        if key < (*node).key {
            (*node).left = insert_rust((*node).left as *mut u8, key) as *mut Node;
        } else {
            (*node).right = insert_rust((*node).right as *mut u8, key) as *mut Node;
        }

        // Return the (unchanged) node pointer
        node as *mut u8
    }
    
}

#[cfg(kani)]
#[kani::proof]
fn main() {
    // Write code to call insert_rust
    let mut root: *mut Node = ptr::null_mut();
    let _result = unsafe { insert_rust(root as *mut u8, 10) };
}
